File: app/src/main/java/com/microsoft/msftbreweryworkshop/service/ServiceLocatorImpl.java
package com.microsoft.msftbreweryworkshop.service;

import android.app.Application;
import com.microsoft.msftbreweryworkshop.BuildConfig;
import com.microsoft.msftbreweryworkshop.api.BeerApi;
import com.microsoft.msftbreweryworkshop.api.BreweryApi;
import com.microsoft.msftbreweryworkshop.network.AuthRequestInterceptor;
import com.microsoft.msftbreweryworkshop.network.HttpClient;
import com.squareup.moshi.Moshi;
import retrofit2.Retrofit;

public class ServiceLocatorImpl implements ServiceLocator {
    private Retrofit retrofit;
    private Application app;

    public ServiceLocatorImpl(Application application) {
        app = application;
        retrofit = HttpClient.buildRetrofit(
                BuildConfig.API_URL,
                HttpClient.buildOkHttpClient(app.getCacheDir()),
                new AuthRequestInterceptor(),
                new Moshi.Builder().build()
        );
    }

    @Override
    public BreweryService getBreweryService() {
        BreweryApi breweryApi = retrofit.create(BreweryApi.class);
        return new BreweryServiceImpl(breweryApi);
    }

    @Override
    public BeerService getBeerService() {
        BeerApi beerApi = retrofit.create(BeerApi.class);
        return new BeerServiceImpl(beerApi);
    }
}

File: app/src/main/java/com/microsoft/msftbreweryworkshop/service/ServiceLocatorImpl.kt
package com.microsoft.msftbreweryworkshop.service

import android.app.Application
import com.microsoft.msftbreweryworkshop.BuildConfig
import com.microsoft.msftbreweryworkshop.api.BeerApi
import com.microsoft.msftbreweryworkshop.api.BreweryApi
import com.microsoft.msftbreweryworkshop.network.AuthRequestInterceptor
import com.microsoft.msftbreweryworkshop.network.buildOkHttpClient
import com.microsoft.msftbreweryworkshop.network.buildRetrofit
import com.squareup.moshi.Moshi
import retrofit2.Retrofit

class ServiceLocatorImpl(val app: Application) : ServiceLocator {
    private val retrofit: Retrofit by lazy {
        buildRetrofit(
            BuildConfig.API_URL,
            buildOkHttpClient(app.cacheDir),
            AuthRequestInterceptor(),
            moshi = Moshi.Builder().build()
        )
    }

    override fun getBreweryService(): BreweryService {
        val breweryApi = retrofit.create(BreweryApi::class.java)
        return BreweryServiceImpl(breweryApi)
    }

    override fun getBeerService(): BeerService {
        val beerApi = retrofit.create(BeerApi::class.java)
        return BeerServiceImpl(beerApi)
    }
}